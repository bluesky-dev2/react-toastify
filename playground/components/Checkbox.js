import React from 'react';

const Checkbox = ({ label, onChange, id, checked }) => (
  <label htmlFor={id}>
    <input
      id={id}
      type="checkbox"
      name={id}
      checked={checked}
      onChange={onChange}
    />
    {label}
  </label>
);

export default Checkbox;
