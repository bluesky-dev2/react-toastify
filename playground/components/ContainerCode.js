import React from 'react';

function getProp(prop, value) {
  const P = () => <span className="code__props">{prop}</span>;
  return value ? (
    <div>
      <span className="code__props">{prop}</span>
    </div>
  ) : (
    <div>
      <span className="code__props">{prop}</span>
      {`={false}`}
    </div>
  );
}

const ContainerCode = ({
  position,
  disableAutoClose,
  autoClose,
  hideProgressBar,
  newestOnTop,
  closeOnClick,
  pauseOnHover,
  rtl,
  pauseOnVisibilityChange,
  isDefaultProps,
  draggable
}) => (
  <div>
    <h3>Toast Container</h3>
    <div className="code">
      <div>
        <span>{`<`}</span>
        <span className="code__component">ToastContainer</span>
      </div>
      <div>
        <span className="code__props">position</span>
        {`="${position}"`}
      </div>
      <div>
        <span className="code__props">autoClose</span>
        {`={${disableAutoClose ? false : autoClose}}`}
      </div>
      {!disableAutoClose ? getProp('hideProgressBar', hideProgressBar) : ''}
      {getProp('newestOnTop', newestOnTop)}
      {getProp('closeOnClick', closeOnClick)}
      {getProp('rtl', rtl)}
      {getProp('pauseOnVisibilityChange', pauseOnVisibilityChange)}
      {getProp('draggable', draggable)}
      {!disableAutoClose ? getProp('pauseOnHover', pauseOnHover) : ''}
      <div>
        <span>{`/>`}</span>
      </div>
      {isDefaultProps && (
        <div>
          <div>{`{/* Same as */}`}</div>
          <span>{`<`}</span>
          <span className="code__component">ToastContainer</span>
          <span> /></span>
        </div>
      )}
    </div>
  </div>
);

export default ContainerCode;
