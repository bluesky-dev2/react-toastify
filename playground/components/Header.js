import React from 'react';

export default () => (
  <header>
    <h1>Welcome to React-toastify</h1>
    <h5>React notification made easy !</h5>
    <div>
      <a className="btn" href="https://github.com/fkhadra/react-toastify/stargazers">
        <span role="img" aria-label="link to github">
          ‚≠êÔ∏è
        </span>
        {' '}Become a stargazer
      </a>
      <a className="btn" href="https://github.com/fkhadra/react-toastify">
        <span role="img" aria-label="link to github">
          üò∏
        </span>
        {' '}View on GitHub
      </a>
    </div>
  </header>
);
