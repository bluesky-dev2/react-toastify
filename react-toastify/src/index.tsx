export { useToastContainer } from './hooks';
export { cssTransition } from './utils';
export { ToastContainer, Bounce, Flip, Slide, Zoom } from './components';
export { toast } from './core';
export * from "./types"