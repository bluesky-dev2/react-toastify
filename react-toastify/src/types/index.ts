import * as React from 'react';

export type ToastPosition =
  | 'top-right'
  | 'top-center'
  | 'top-left'
  | 'bottom-right'
  | 'bottom-center'
  | 'bottom-left';

export type TypeOptions = 'info' | 'success' | 'warning' | 'error' | 'default';
export type ToastContent = React.ReactNode | (() => React.ReactNode );
export interface Toast {
  content: ToastContent;
  options: InternalToastOptions;
}

type Id = number | string
export type ToastId = Id;
export type ContainerId = Id;

interface CommonOptions {
  /**
   * Pause the timer when the mouse hover the toast.
   * `Default: true`
   */
  pauseOnHover?: boolean;

  /**
   * Pause the toast when the window loose focus.
   * `Default: true`
   */
  pauseOnFocusLoss?: boolean;

  /**
   * Remove the toast when clicked.
   * `Default: true`
   */
  closeOnClick?: boolean;

  /**
   * Set the delay in ms to close the toast automatically.
   * Use `false` to prevent the toast from closing.
   * `Default: 5000`
   */
  autoClose?: number | false;

  /**
   * Set the default position to use.
   * `One of: 'top-right', 'top-center', 'top-left', 'bottom-right', 'bottom-center', 'bottom-left'`
   * `Default: 'top-right'`
   */
  position?: ToastPosition;

  /**
   * Pass a custom close button.
   * To remove the close button pass `false`
   */
  closeButton?: React.ReactNode | boolean;

  /**
   * An optional css class to set for the progress bar.
   */
  progressClassName?: string | object | null;

  /**
   * An optional style to set for the progress bar.
   */
  progressStyle?: object;

  /**
   * An optional css class to set.
   */
  className?: string | object | null;

  /**
   * An optional css class to set for the toast content.
   */
  bodyClassName?: string | object | null;

  /**
   * Hide or show the progress bar.
   * `Default: false`
   */
  hideProgressBar?: boolean;

  /**
   * Pass a custom transition built with react-transition-group.
   */
  transition?: React.ComponentType;

  /**
   * Allow toast to be draggable
   * `Default: true`
   */
  draggable?: boolean;

  /**
   * The percentage of the toast's width it takes for a drag to dismiss a toast
   * `Default: 80`
   */
  draggablePercent?: number;

  /**
   * Define the ARIA role for the toast
   * `Default: alert`
   *  https://www.w3.org/WAI/PF/aria/roles
   */
  role?: string;

  /**
   * Set id to handle multiple container
   */
  containerId?: ContainerId;

  /**
   * Fired when clicking inside toaster
   */
  onClick?: Function;

  /**
   * Support right to left display.
   * `Default: false`
   */
  rtl?: boolean;
}

export interface ToastOptions extends CommonOptions {
  /**
   * Called when toast is mounted.
   */
  onOpen?: () => void;

  /**
   * Called when toast is unmounted.
   */
  onClose?: () => void;

  /**
   * Set the toast type.
   * `One of: 'info', 'success', 'warning', 'error', 'default'`
   */
  type?: TypeOptions;

  /**
   * Set a custom `toastId`
   */
  toastId?: ToastId;

  /**
   * Used during update
   */
  updateId?: ToastId;

  /**
   * Set the percentage for the controlled progress bar. `Value must be between 0 and 1.`
   */
  progress?: number | string;

  /**
   * Add a delay in ms before the toast appear.
   */
  delay?: number;
}

export interface InternalToastOptions extends ToastOptions {
  staleId?: ToastId;
  toastId: ToastId;
  key: Id;
  closeToast: () => void;
  position: ToastPosition;
}

export interface UpdateOptions extends ToastOptions {
  /**
   * Used to update a toast.
   * Pass any valid ReactNode(string, number, component)
   */
  render?: ToastContent;
}

export interface ToastContainerProps extends CommonOptions {
  /**
   * Whether or not to display the newest toast on top.
   * `Default: false`
   */
  newestOnTop?: boolean;

  /**
   * An optional inline style to apply.
   */
  style?: object;

  /**
   * An optional css class for the toast.
   */
  toastClassName?: string | object | null;

  /**
   * Show the toast only if it includes containerId and it's the same as containerId
   * `Default: false`
   */
  enableMultiContainer?: boolean;
}

// interface Toast {
//   /**
//    * Shorthand to display toast of type 'success'.
//    */
//   success(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'info'.
//    */
//   info(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'warning'.
//    */
//   warn(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Shorthand to display toast of type 'error'.
//    */
//   error(content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Check if a toast is active by passing the `toastId`.
//    * Each time you display a toast you receive a `toastId`.
//    */
//   isActive(toastId: ToastId): boolean;

//   /**
//    * Remove a toast. If no `toastId` is used, all the active toast
//    * will be removed.
//    */
//   dismiss(toastId?: ToastId): void;

//   /**
//    * Update an existing toast. By default, we keep the initial content and options of the toast.
//    */
//   update(toastId: ToastId, options?: UpdateOptions): void;

//   /**
//    * Listen for change when a toast is added or removed. The number of toast displayed is passed as paran to the callback
//    */
//   onChange(
//     callback: (count?: number, containerId?: string | number) => void
//   ): void;

//   /**
//    * Set a controlled progress bar value to 100% then close the toast
//    */
//   done(toastId: ToastId): void;

//   /**
//    * Let you define `ToastContainer` props when lazy mounted.
//    * When called enable lazy mounted container
//    */
//   configure(config?: ToastContainerProps): void;

//   /**
//    * Display a toast without a specific type.
//    */
//   (content: ToastContent, options?: ToastOptions): ToastId;

//   /**
//    * Helper to set notification type
//    */
//   TYPE: Type;

//   /**
//    * Helper to set position
//    */
//   POSITION: Position;
// }

// /**
//  * Helper to set notification type
//  */

// /**
//  * Helper to set position
//  */

// /**
//  * Helper to build custom entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */

// /**
//  * Built-in entrance and exit transition
//  */
