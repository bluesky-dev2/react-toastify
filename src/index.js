import ToastContainer from './ToastContainer';
import toastify from './toastify';

export { ToastContainer, toastify as toast };
