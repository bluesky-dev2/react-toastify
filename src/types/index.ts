import * as React from 'react';

export type ToastPosition =
  | 'top-right'
  | 'top-center'
  | 'top-left'
  | 'bottom-right'
  | 'bottom-center'
  | 'bottom-left';

export type TypeOptions = 'info' | 'success' | 'warning' | 'error' | 'default';
export type ToastContent = React.ReactNode | ((props?: any) => React.ReactNode);
export interface Toast {
  content: ToastContent;
  options: WithInjectedOptions;
}

type Id = number | string;
type ClassName = string | null;
export type ToastId = Id;
export type ContainerId = Id;

interface CommonOptions {
  /**
   * Pause the timer when the mouse hover the toast.
   * `Default: true`
   */
  pauseOnHover?: boolean;

  /**
   * Pause the toast when the window loose focus.
   * `Default: true`
   */
  pauseOnFocusLoss?: boolean;

  /**
   * Remove the toast when clicked.
   * `Default: true`
   */
  closeOnClick?: boolean;

  /**
   * Set the delay in ms to close the toast automatically.
   * Use `false` to prevent the toast from closing.
   * `Default: 5000`
   */
  autoClose?: number | false;

  /**
   * Set the default position to use.
   * `One of: 'top-right', 'top-center', 'top-left', 'bottom-right', 'bottom-center', 'bottom-left'`
   * `Default: 'top-right'`
   */
  position?: ToastPosition;

  /**
   * Pass a custom close button.
   * To remove the close button pass `false`
   */
  closeButton?:
    | React.ReactElement
    | ((props: any) => React.ReactElement)
    | boolean;

  /**
   * An optional css class to set for the progress bar.
   */
  progressClassName?: ClassName;

  /**
   * An optional style to set for the progress bar.
   */
  progressStyle?: React.CSSProperties;

  /**
   * An optional css class to set.
   */
  className?: ClassName;

  /**
   * An optional css class to set for the toast content.
   */
  bodyClassName?: ClassName;

  /**
   * An optional inline style to apply for the toast content.
   */
  bodyStyle?: React.CSSProperties;

  /**
   * Hide or show the progress bar.
   * `Default: false`
   */
  hideProgressBar?: boolean;

  /**
   * Pass a custom transition built with react-transition-group.
   */
  transition?: React.ReactNode;

  /**
   * Allow toast to be draggable
   * `Default: true`
   */
  draggable?: boolean;

  /**
   * The percentage of the toast's width it takes for a drag to dismiss a toast
   * `Default: 80`
   */
  draggablePercent?: number;

  /**
   * Define the ARIA role for the toast
   * `Default: alert`
   *  https://www.w3.org/WAI/PF/aria/roles
   */
  role?: string;

  /**
   * Set id to handle multiple container
   */
  containerId?: ContainerId;

  /**
   * Fired when clicking inside toaster
   */
  onClick?: (event: React.MouseEvent) => void;

  /**
   * Support right to left display.
   * `Default: false`
   */
  rtl?: boolean;
}

export interface ToastOptions extends CommonOptions {
  /**
   * Called when toast is mounted.
   */
  onOpen?: <T = {}>(props: T) => void;

  /**
   * Called when toast is unmounted.
   */
  onClose?: <T = {}>(props: T) => void;

  /**
   * An optional inline style to apply.
   */
  style?: React.CSSProperties;

  /**
   * Set the toast type.
   * `One of: 'info', 'success', 'warning', 'error', 'default'`
   */
  type?: TypeOptions;

  /**
   * Set a custom `toastId`
   */
  toastId?: ToastId;

  /**
   * Used during update
   */
  updateId?: ToastId;

  /**
   * Set the percentage for the controlled progress bar. `Value must be between 0 and 1.`
   */
  progress?: number | string;

  /**
   * Add a delay in ms before the toast appear.
   */
  delay?: number;
}

export interface WithInjectedOptions extends ToastOptions {
  staleId?: ToastId;
  toastId: ToastId;
  key: Id;
  closeToast: () => void;
  position: ToastPosition;
  children?: ToastContent;
  draggablePercent: number;
  in?: boolean;
  progressClassName?: ClassName;
  className?: ClassName;
  bodyClassName?: ClassName;
  onExited?: () => void;
  // transition: Transition;
}

export interface UpdateOptions extends ToastOptions {
  /**
   * Used to update a toast.
   * Pass any valid ReactNode(string, number, component)
   */
  render?: ToastContent;
}

export interface ToastContainerProps extends CommonOptions {
  /**
   * Whether or not to display the newest toast on top.
   * `Default: false`
   */
  newestOnTop?: boolean;

  /**
   * An optional inline style to apply.
   */
  style?: React.CSSProperties;

  /**
   * An optional inline style to apply for the toast.
   */
  toastStyle?: React.CSSProperties;

  /**
   * An optional css class for the toast.
   */
  toastClassName?: ClassName;

  /**
   * Show the toast only if it includes containerId and it's the same as containerId
   * `Default: false`
   */
  enableMultiContainer?: boolean;

  /**
   * Limit the number of toast displayed at the same time
   */
  limit?: number;
}
