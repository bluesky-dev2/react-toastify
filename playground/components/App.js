import React, { Component } from 'react';

import Header from './Header';
import Radio from './Radio';
import Checkbox from './Checkbox';
import ContainerCode from './ContainerCode';
import ToastCode from './ToastCode';

import { ToastContainer, toast } from './../../src/index';

const flags = [
  {
    id: 'disableAutoClose',
    label: 'Disable auto-close'
  },
  {
    id: 'hideProgressBar',
    label: 'Hide progress bar(less fanciness!)'
  },
  {
    id: 'newestOnTop',
    label: 'Newest on top*'
  },
  {
    id: 'closeOnClick',
    label: 'Close on click'
  },
  {
    id: 'pauseOnHover',
    label: 'Pause delay on hover'
  },
  {
    id: 'rtl',
    label: 'Right to left layout*'
  },
  {
    id: 'pauseOnVisibilityChange',
    label: 'Pause toast on visibility change*'
  }
];

class App extends Component {
  state = App.getDefaultState();

  static getDefaultState() {
    return {
      ...ToastContainer.defaultProps,
      type: 'default',
      disableAutoClose: false
    };
  }

  handleReset = () =>
    this.setState({
      ...App.getDefaultState()
    });

  clearAll = () => toast.dismiss();

  // showToast = () => toast

  handleAutoCloseDelay = e =>
    this.setState({
      autoClose: e.target.value > 0 ? parseInt(e.target.value, 10) : 1
    });

  handleRadio = e =>
    this.setState({
      [e.target.name]: e.target.value
    });

  toggleCheckbox = e =>
    this.setState({
      [e.target.name]: !this.state[e.target.name]
    });

  renderFlags() {
    return flags.map(({ id, label }) => (
      <li key={id}>
        <Checkbox
          id={id}
          label={label}
          onChange={this.toggleCheckbox}
          checked={this.state[id]}
        />
      </li>
    ));
  }

  render() {
    return (
      <main>
        <Header />
        <div className="container">
            <p>By default, all toasts will inherit ToastContainer's props. Props defined on toast supersede ToastContainer's props.
              Props marked with * can only be set on the ToastContainer.
            </p>
          <section className="container__options">
            <div>
              <h3>Position</h3>
              <ul>
                <Radio
                  options={toast.POSITION}
                  name="position"
                  checked={this.state.position}
                  onChange={this.handleRadio}
                />
              </ul>
            </div>
            <div>
              <h3>Type</h3>
              <ul>
                <Radio
                  options={toast.TYPE}
                  name="type"
                  checked={this.state.type}
                  onChange={this.handleRadio}
                />
              </ul>
            </div>
            <div>
              <h3>Options</h3>
              <label htmlFor="autoClose">
                Delay
                <input
                  type="number"
                  name="autoClose"
                  id="autoClose"
                  value={this.state.autoClose}
                  onChange={this.handleAutoCloseDelay}
                  disabled={this.state.disableAutoClose}
                />
                ms
              </label>
              <ul>{this.renderFlags()}</ul>
              <ul className="container__actions">
                <li>
                  <button className="btn">
                    <span role="img" aria-label="show alert">
                      ðŸš€
                    </span>{' '}
                    Show Toast
                  </button>
                </li>
                <li>
                  <button className="btn bg-red" onClick={this.clearAll}>
                    <span role="img" aria-label="clear all">
                      ðŸ’©
                    </span>{' '}
                    Clear All
                  </button>
                </li>
                <li>
                  <button className="btn bg-blue" onClick={this.handleReset}>
                    <span role="img" aria-label="reset options">
                      ðŸ”„
                    </span>{' '}
                    Reset
                  </button>
                </li>
              </ul>
            </div>
          </section>
          <section>
            <ContainerCode {...this.state} />
            <ToastCode {...this.state} />
          </section>
        </div>
        <ToastContainer />
      </main>
    );
  }
}

export default App;
