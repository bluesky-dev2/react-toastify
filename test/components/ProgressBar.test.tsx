import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { ProgressBar } from '../../src/components';
import { Theme } from '../../src/types';

const getProps = () => ({
  delay: 5000,
  isRunning: true,
  rtl: false,
  closeToast: jest.fn(),
  isIn: true,
  theme: ['colored', 'light', 'dark'][Math.floor(Math.random() * 3)] as Theme
});

describe('ProgressBar', () => {
  it('Should merge className', () => {
    render(<ProgressBar {...getProps()} className="test" />);

    expect(document.querySelector('.test')?.classList.contains('test')).toBe(
      true
    );
  });
  it('Should merge className in function form', () => {
    render(<ProgressBar {...getProps()} className={() => 'testClassName'} />);

    expect(
      document
        .querySelector('.testClassName')
        ?.classList.contains('testClassName')
    ).toBe(true);
  });
  it('Should call closeToast function when animation end', () => {
    const closeToast = jest.fn();
    const { container } = render(
      <ProgressBar {...getProps()} closeToast={closeToast} />
    );

    expect(closeToast).not.toHaveBeenCalled();
    fireEvent.animationEnd(container.firstChild as HTMLElement);
    expect(closeToast).toHaveBeenCalled();
  });

  it('Should be able to hide the progress bar', () => {
    const { container } = render(<ProgressBar {...getProps()} hide />);
    expect((container.firstChild! as HTMLElement).style.opacity).toBe('0');
  });

  it('Should be able to pause animation', () => {
    const { container } = render(
      <ProgressBar {...getProps()} isRunning={false} />
    );
    expect(
      (container.firstChild! as HTMLElement).style.animationPlayState
    ).toBe('paused');
  });

  it('Should render controlled progress bar', () => {
    const { container } = render(
      <ProgressBar {...getProps()} controlledProgress progress={0.7} />
    );
    expect((container.firstChild! as HTMLElement).style.transform).toBe(
      'scaleX(0.7)'
    );
  });
});
