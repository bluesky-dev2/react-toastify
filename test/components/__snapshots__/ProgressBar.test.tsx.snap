// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar Should be able to hide the progress bar 1`] = `
<div>
  <div
    aria-hidden="true"
    aria-label="notification timer"
    class="Toastify__progress-bar Toastify__progress-bar--animated Toastify__progress-bar--default"
    role="progressbar"
    style="animation-duration: 5000ms; animation-play-state: running; opacity: 0;"
  />
</div>
`;

exports[`ProgressBar Should be able to pause animation 1`] = `
<div>
  <div
    aria-hidden="false"
    aria-label="notification timer"
    class="Toastify__progress-bar Toastify__progress-bar--animated Toastify__progress-bar--default"
    role="progressbar"
    style="animation-duration: 5000ms; animation-play-state: paused; opacity: 1;"
  />
</div>
`;

exports[`ProgressBar Should render controlled progress bar 1`] = `
<div>
  <div
    aria-hidden="false"
    aria-label="notification timer"
    class="Toastify__progress-bar Toastify__progress-bar--controlled Toastify__progress-bar--default"
    role="progressbar"
    style="animation-duration: 5000ms; animation-play-state: running; opacity: 1; transform: scaleX(0.7);"
  />
</div>
`;
